### DISCUSSION TITLE
ModuleNotFoundError: No module named 'pymodbus.client.sync'

### DISCUSSION THREAD

**Discussion by smmadge** [addiejanssen/domoticz-solaredge-modbustcp-plugin#62] [Q&A] [answered]
Trying to install the solaredge module but somehow it doesn't work. I did a sudo 'pip3 install -r requirements.txt' and it installed solaredge-modbus 0.7.0 succesfully but it won't run. I see the following in the debug log:

Mar 17 16:07:34 iot domoticz: SolarEdge SE3000H: (SolarEdge_ModbusTCP) failed to load 'plugin.py', Python Path used was '/home/domoticz/domoticz/plugins/domoticz-solaredge-modbustcp-plugin/:/usr/lib/python310.zip:/usr/lib/python3.10:/usr/lib/python3.10/lib-dynload:/usr/local/lib/python3.10/dist-packages:/usr/lib/python3/dist-packages:/usr/lib/python3.10/dist-packages'.
Mar 17 16:07:34 iot domoticz: SolarEdge SE3000H: Traceback (most recent call last):
Mar 17 16:07:34 iot domoticz: SolarEdge SE3000H:   File "/home/domoticz/domoticz/plugins/domoticz-solaredge-modbustcp-plugin/plugin.py", line 54, in <module>
Mar 17 16:07:34 iot domoticz: SolarEdge SE3000H:     import solaredge_modbus
Mar 17 16:07:34 iot domoticz: SolarEdge SE3000H:   File "/usr/local/lib/python3.10/dist-packages/solaredge_modbus/__init__.py", line 7, in <module>
Mar 17 16:07:34 iot domoticz: SolarEdge SE3000H:     from pymodbus.client.sync import ModbusTcpClient
Mar 17 16:07:34 iot domoticz: SolarEdge SE3000H: ModuleNotFoundError: No module named 'pymodbus.client.sync'

Initially i tried with pymodbus 3.2.0 (latest) but also tried with pymodbus=3.0.0

Any idea what i am doing wrong?

**Comment by smmadge** [comment, marked as answer]
Fixed, apparently using the following (at the time of writing latest) packages fixed it:
solaredge_modbus-0.7.2-py3-none-any.whl 
pyserial_asyncio-0.6-py3-none-any.whl 
pymodbus-3.2.1-py3-none-any.whl 

First uninstall all previously installed python modules:
sudo pip3 uninstall solaredge_modbus
sudo pip3 uninstall pyserial_asyncio
sudo pip3 uninstall pymodbus

and then do a
sudo pip3 install solaredge_modbus

In contrary of what the README.txt says: do NOT use the requirements.txt dependency since that would install an older solaredge_modbus version which doesn't seem to work well with pymodbus anymore.

**Comment by mdklapwijk** [comment]
Had an issue getting the plugin v1.1.1 to start in the docker image domoticz/domoticz:beta (version 2023.2#15615); it also failed with:
```
2023-10-25 23:07:53.473 Error: se6000h: (SolarEdge_ModbusTCP) failed to load 'plugin.py', Python Path used was '/opt/domoticz/userdata/plugins/modbustcp/:/usr/lib/python39.zip:/usr/lib/python3.9:/usr/lib/python3.9/lib-dynload:/usr/local/lib/python3.9/dist-packages:/usr/lib/python3/dist-packages:/usr/lib/python3.9/dist-packages'.
2023-10-25 23:07:53.474 Error: se6000h: Traceback (most recent call last):
2023-10-25 23:07:53.474 Error: se6000h: File "/opt/domoticz/userdata/plugins/modbustcp/plugin.py", line 54, in <module>
2023-10-25 23:07:53.474 Error: se6000h: import solaredge_modbus
2023-10-25 23:07:53.474 Error: se6000h: File "/usr/local/lib/python3.9/dist-packages/solaredge_modbus/__init__.py", line 7, in <module>
2023-10-25 23:07:53.474 Error: se6000h: from pymodbus.client.sync import ModbusTcpClient
2023-10-25 23:07:53.475 Error: se6000h: ModuleNotFoundError: No module named 'pymodbus.client.sync'
```

Setting the following in the /opt/domoticz/userdata/customstart.sh fixed it for me:
```
#!/bin/bash
pip install requests
pip3 install requests
pip3 install pyserial_asyncio==0.6
pip3 install pymodbus==2.5.3
pip3 install solaredge_modbus==0.7.0
```
note: pymodbus version seems key in this and requests is needed for the deconz plugin I am using.

**Reply by telewy** [reply]
I jave just gave up with this plugin. Spen few hours trying to get it working on my new Debian 12 server. Finally I gave up i went back to my original approach which is python script reading data from inverter and sending it to domoticz using mqtt. I am not programmer but preparin scrip was pretty easy. I addapted example.py script from https://github.com/nmakel/solaredge_modbus. Have the same functionalities like with plugin just had to create dummy devices by myself.
